FEAT: Basic setup

matrixor411 opened the pull request:
- Jest + sample test,
- Prisma + scheme,
- basic endpoints
- env + basic config

matrixor411 committed:
FEAT: Basic setup
- Jest + sample test,
- Prisma + scheme,
- basic endpoints
- env + basic config

shadowfiga commented on prisma/migrations/20231208150650_init/migration.sql:
```diff
@@ -0,0 +1,15 @@
+    "id" SERIAL NOT NULL,
+    "email" TEXT NOT NULL,
+    "password" TEXT NOT NULL,
+    "role" "Role" NOT NULL DEFAULT 'USER',
```
Prisma schema and migrations do not match, please regenerate the migrations.

shadowfiga commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,17 @@
+model User {
+  id        Int         @id @default(autoincrement())
+  email  String      @unique
+  password  String
```
Maybe a better name could be passwordDigest or passwordHash.

shadowfiga commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,17 @@
+}
+
+model User {
+  id        Int         @id @default(autoincrement())
```
This should be a uuid.

shadowfiga commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,17 @@
+  url      = env("DATABASE_URL")
+}
+
+model User {
```
(Personal preference) Naming this something other than user might be better, up to you, it's fine by me either way.

shadowfiga commented on src/config/configuration.ts:
```diff
@@ -0,0 +1,3 @@
+export default () => ({
+  server_port: parseInt(process.env.PORT, 10) || 3000,
```
Inconsistent naming, use camelCase.

shadowfiga commented on src/app.module.ts:
```diff
@@ -1,9 +1,18 @@
 @Module({
-  imports: [],
+  imports: [
+    ConfigModule.forRoot({
```
Config modules can be global, please look at the docs and make this a global menu.

shadowfiga commented on src/modules/user/tests/user.controller.spec.ts:
```diff
@@ -0,0 +1,64 @@
+import { User } from '../user.entity';
+import { NotFoundException, UnauthorizedException } from '@nestjs/common';
+
+describe('UserService', () => {
```
Please move this to a file named user.controller.e2e.ts, sine these are e2e tests.

shadowfiga commented on src/modules/user/user.controller.ts:
```diff
@@ -0,0 +1,41 @@
+
+  @Post()
+  async login(
+    @Body('email') email: string,
```
Use class-validator, does this validate the request? Please use this [class-validator](https://github.com/typestack/class-validator).

shadowfiga commented on src/modules/user/user.controller.ts:
```diff
@@ -0,0 +1,41 @@
+    @Body('email') email: string,
+    @Body('password') password: string,
+  ) {
+    try {
```
No try or catch allowed.

shadowfiga replied:
Please take a look at global exception pipes, etc.

shadowfiga commented on src/modules/user/user.controller.ts:
```diff
@@ -0,0 +1,41 @@
+    @Param('id') id: number,
+    @Body('password') newPassword: string,
+  ) {
+    return this.userService.updatePassword(id, newPassword);
```
Where is the currently logged in user being validated? Why would I need to pass in id for the user if I am validated as the user itself? Please take a look at passport and how to properly handle authentication and authorization.

shadowfiga commented on src/modules/user/user.entity.ts:
```diff
@@ -0,0 +1,5 @@
+export class User {
```
Why would you need to export an entity? Prisma client automatically generates the appropriate types. Please take a look at the prisma/client documentation.

shadowfiga commented on src/modules/user/user.repository.ts:
```diff
@@ -0,0 +1,24 @@
+import { PrismaClient, User } from '@prisma/client';
+const prisma = new PrismaClient();
```
Please look at the proper guide on how to integrate prisma with nestjs, this is not the proper way, the service itself should be injectable and available in a global module.

shadowfiga commented on src/modules/user/user.repository.ts:
```diff
@@ -0,0 +1,24 @@
+import { PrismaClient, User } from '@prisma/client';
+const prisma = new PrismaClient();
+export interface UserRepository {
```
Remove, no need for definitions if there is only one way of doing concrete implementation, we don't plan on doing a v2 any time soon.

shadowfiga commented on src/modules/user/user.repository.ts:
```diff
@@ -0,0 +1,24 @@
+  updatePassword(id: number, newPassword: string): Promise<User | null>;
+}
+
+export class UserRepositoryImpl implements UserRepository {
```
Missing @Injectible() decorator, when using prisma as the repository "provider", the repository itself is a service, so please fix this.

shadowfiga commented on src/modules/user/user.module.ts:
```diff
@@ -0,0 +1,13 @@
+  controllers: [UserController],
+  providers: [
+    UserService,
+    { provide: 'UserRepository', useClass: UserRepositoryImpl },
```
This should only be UserRepository, no need for abstractions at this point. Make sure the "repository" is injectable.

shadowfiga commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,38 @@
+  async login(email: string, password: string): Promise<User | null> {
+    const user = await this.userRepository.findByEmail(email);
+
+    if (!user) {
```
Prisma client should expose a function called findOneOrThrow or something along those lines, please use that one, no need to reinvent the wheel.

shadowfiga commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,38 @@
+    }
+    const isPasswordValid = user.password === password;
+
+    if (!isPasswordValid) {
```
It's never a good idea to throw credential specific errors, this opens up applications to brute force attacks and the attacker now knows that there is a user with a given email for which the password is missing.

shadowfiga commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,38 @@
+    if (!user) {
+      throw new NotFoundException(`User with username ${email} not found`);
+    }
+    const isPasswordValid = user.password === password;
```
Please integrate bcrypt or some other library to hash and compare the passwords properly.

shadowfiga commented on src/modules/user/user.service.ts:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  async updatePassword(id: number, newPassword: string): Promise<User | null> {
+    const user = await this.userRepository.updatePassword(id, newPassword);
```
The repository should be responsible for throwing this error, now imagine we have to update the user in 3 different controllers, this results in 3x repeating this if statement. If it's the repository that takes care of it, only the repository will throw without any code repetition. Semantically it also makes more sense for the repo to throw.

shadowfiga commented on test/app.e2e-spec.ts:
```diff
@@ -1,7 +1,7 @@
 import { INestApplication } from '@nestjs/common';
 import * as request from 'supertest';
-import { AppModule } from './../src/app.module';
+import { AppModule } from '../src/app.module';
```
Prefer relative imports, please take a look at your IDE settings and tsconfig.json. You shoul be able to import this module by writting.
```ts
import {AppModule} from 'src/app.module'
```

shadowfiga reviewed (changes requested):
Good first pass, please fix all of the issues that I brought up 👍

matrixor411 committed:
Changes based on latest pr

shadowfiga commented on src/config/configuration.ts:
```diff
@@ -0,0 +1,4 @@
+export default () => ({
```
Missing validation.

shadowfiga commented on src/modules/account/account.controller.ts:
```diff
@@ -0,0 +1,55 @@
+      loginInput.email,
+      loginInput.password,
+    );
+    response.cookie('access_token', jwtPayload.access_token);
```
Create a global enum for storage related things.

shadowfiga replied:
Inconsistent naming strategy, please fix.

shadowfiga commented on src/modules/account/account.repository.ts:
```diff
@@ -0,0 +1,25 @@
+    newPassword: string,
+  ): Promise<Account | null> {
+    const existingAccount = await this.prisma.account.update({
+      where: { account_id: accountId },
```
Inconsistent naming strategy.

shadowfiga commented on src/modules/account/account.service.ts:
```diff
@@ -0,0 +1,30 @@
+    const account = await this.accountRepository.findByEmail(email);
+    await comparePasswords(password, account.hashedPassword);
+
+    const payload = { sub: account.account_id, email: account.email };
```
Remove email and create an internal identifier.

shadowfiga commented on src/util/bcrypt/bcrypt.compare.ts:
```diff
@@ -0,0 +1,17 @@
+  plainPassword: string,
+  hashedPassword: string,
+): Promise<void> => {
+  await bcrypt.compare(plainPassword, hashedPassword).then(function (
```
I think that compare throws by default so you can actually get rid of everything.

shadowfiga commented on src/modules/account/auth/jwt.strategy.ts:
```diff
@@ -0,0 +1,36 @@
+
+@Injectable()
+export class JwtStrategy extends PassportStrategy(Strategy) {
+  constructor() {
```
Consume ConfigService.

shadowfiga commented on src/modules/account/auth/jwt.strategy.ts:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  async validate(payload: any) {
+    return { userId: payload.sub, username: payload.username };
```
Validate if the account actually exists.

shadowfiga commented on src/modules/account/auth/local.strategy.ts:
```diff
@@ -0,0 +1,23 @@
+import { PassportStrategy } from '@nestjs/passport';
```
Remove this strategy.

shadowfiga commented on src/modules/account/dto/login.response.ts:
Remove the empty file.

shadowfiga commented on src/util/date/date.specific-dates.ts:
```diff
@@ -0,0 +1 @@
+export const dateNow = new Date(Date.now());
```
Remove.

shadowfiga commented on src/util/filter/fliter.notFoundException.ts:
```diff
@@ -0,0 +1,27 @@
+    const ctx = host.switchToHttp();
+    const response = ctx.getResponse<Response>();
+
+    switch (exception.code) {
```
Use if's instead for better readability.

shadowfiga commented on src/util/filter/fliter.notFoundException.ts:
```diff
@@ -0,0 +1,27 @@
+@Catch(Prisma.PrismaClientKnownRequestError)
+export class PrismaClientExceptionFilter extends BaseExceptionFilter {
+  catch(exception: Prisma.PrismaClientKnownRequestError, host: ArgumentsHost) {
+    console.error(exception.message);
```
No consoles!

shadowfiga reviewed (changes requested):
A little bit more.

matrixor411 committed:
FEAT: Changes based on pr rereview

matrixor411 committed:
FEAT: changed Config validation method

shadowfiga commented on src/modules/account/account.controller.ts:
```diff
@@ -0,0 +1,44 @@
+    @Req() request: Request,
+    @Body() updatePasswordInput: UpdatePasswordInput,
+  ) {
+    return this.accountService.updatePassword('', updatePasswordInput.password);
```
You are missing accountId?

shadowfiga commented on src/modules/account/account.repository.ts:
```diff
@@ -0,0 +1,32 @@
+  ): Promise<Account | null> {
+    const existingAccount = await this.prisma.account.update({
+      where: { accountId: accountId },
+      data: { hashedPassword: newPassword },
```
This password seems raw, did you forget to create a hashed version of it?

shadowfiga commented on src/modules/account/auth/jwt.strategy.ts:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  private static extractJwtFromCookie(req: Request): string | null {
+    console.log(req.cookies);
```
No console.log allowed

shadowfiga commented on src/util/constants.ts:
```diff
@@ -0,0 +1,3 @@
+export enum storage {
```
Use PascalCase for type definitions.

shadowfiga commented on src/util/constants.ts:
```diff
@@ -0,0 +1,3 @@
+export enum storage {
+  jwtTokenName = 'accessToken',
```
Enum keys should have the same value as the keys. So what it boils down to is
```typescript
export enum StorageKeysEnum {
  accessToken = "accessToken"
}
```

shadowfiga commented on src/util/filter/fliter.notFoundException.ts:
```diff
@@ -0,0 +1,22 @@
+    const ctx = host.switchToHttp();
+    const response = ctx.getResponse<Response>();
+
+    if (exception.code == 'P2025') {
```
Explain what p2025 means? In this case I would create a one of enum that has the keys named properly and has the specific values associated with it, e.g. 
```typescript
export enum DatabaseExceptionsEnum {
  notFound = 'P2025'
}
```

shadowfiga reviewed (changes requested):
Almost there

matrixor411 committed:
FEAT: Changes based on lastest pr rereview
- Updated reset password endp
- hash password in acc service
- modified expires in jwt
- added enum for notFound
- named our custom passport strategy
- removed console log
